llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -4563,6 +4563,49 @@ static bool isInterleaveShuffle(ArrayRef<int> Mask, MVT VT, int &EvenSrc,
   return ((EvenSrc % HalfNumElts) == 0) && ((OddSrc % HalfNumElts) == 0);
 }
 
+/// Is this mask representing a masked combination of two slides?
+static bool isMaskedSlidePair(ArrayRef<int> Mask,
+                              std::pair<int, int> SrcInfo[2]) {
+  int NumElts = Mask.size();
+  int SIGNAL_VAL = NumElts * 2;
+  SrcInfo[0] = {-1, SIGNAL_VAL};
+  SrcInfo[1] = {-1, SIGNAL_VAL};
+  for (unsigned i = 0; i != Mask.size(); ++i) {
+    int M = Mask[i];
+    if (M < 0)
+      continue;
+    int Src = M >= (int)NumElts;
+    int Diff = (M % NumElts) - (int)i;
+    bool Match = false;
+    for (int j = 0; j < 2; j++) {
+      if (SrcInfo[j].first == -1)
+        SrcInfo[j].first = Src;
+      if (SrcInfo[j].second == SIGNAL_VAL)
+        SrcInfo[j].second = Diff;
+      if (SrcInfo[j].first == Src && SrcInfo[j].second == Diff) {
+        Match = true;
```

Comment:
Would it be easier to compute the mask here where we already know which slide each element is coming from, `j`?

Reply:
I'd rather not.  We could, but then we'd have to integrate the vslideup special case there too.  I'd rather keep the complexity manageable.